File: artifacts/api-server/src/domain/weighted_grade_book.py
"""WeightedGradeBook — standard weighted average grading scheme."""
from typing import List
from .grade_book import GradeBook, AssignmentScore, GradeResult


class WeightedGradeBook(GradeBook):
    @property
    def scheme_name(self) -> str:
        return "weighted"

    def calculate_grade(self, scores: List[AssignmentScore]) -> GradeResult:
        if not scores:
            return GradeResult(percentage=0.0, letter_grade="F", grade_points=0.0, display_label="F")
        pct = self._weighted_percentage(scores)
        letter = self._score_to_letter(pct)
        return GradeResult(
            percentage=round(pct * 10) / 10,
            letter_grade=letter,
            grade_points=self._letter_to_points(letter),
            display_label=letter,
        )

File: artifacts/api-server/src/domain/WeightedGradeBook.ts
import { GradeBook, AssignmentScore, GradeResult } from "./GradeBook";

export class WeightedGradeBook extends GradeBook {
  get schemeName(): string {
    return "weighted";
  }

  calculateGrade(scores: AssignmentScore[]): GradeResult {
    if (scores.length === 0) {
      return { percentage: 0, letterGrade: "F", gradePoints: 0.0, displayLabel: "F" };
    }
    const pct = this.weightedPercentage(scores);
    const letter = this.scoreToLetter(pct);
    return {
      percentage: Math.round(pct * 10) / 10,
      letterGrade: letter,
      gradePoints: this.letterToPoints(letter),
      displayLabel: letter,
    };
  }
}
